lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -0,0 +1,909 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//===----------------------------------------------------------------------===//
+
+#include "circt/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.h"
+#include "circt/Dialect/FIRRTL/Ops.h"
+#include "circt/Dialect/Handshake/HandshakeOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Function.h"
+#include "mlir/IR/OpImplementation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Types.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/Utils.h"
+
+using namespace mlir;
+using namespace mlir::handshake;
+using namespace circt;
+using namespace circt::firrtl;
+
+using ValueVector = std::vector<Value>;
+using ValueVectorList = std::vector<ValueVector>;
+
+//===----------------------------------------------------------------------===//
+// Utils
+//===----------------------------------------------------------------------===//
+
+/// Build a FIRRTL bundle type (with data, valid, and ready subfields) given the
+/// type of the data subfield.
+FIRRTLType buildBundleType(FIRRTLType dataType, bool isFlip,
+                           MLIRContext *context) {
+  using BundleElement = std::pair<Identifier, FIRRTLType>;
+  llvm::SmallVector<BundleElement, 3> elements;
+
+  // Add valid and ready subfield to the bundle.
+  Identifier validId = Identifier::get("valid", context);
+  Identifier readyId = Identifier::get("ready", context);
+  UIntType signalType = UIntType::get(context, 1);
+  if (isFlip) {
+    elements.push_back(BundleElement(validId, FlipType::get(signalType)));
+    elements.push_back(BundleElement(readyId, signalType));
+  } else {
+    elements.push_back(BundleElement(validId, signalType));
+    elements.push_back(BundleElement(readyId, FlipType::get(signalType)));
+  }
+
+  // Add data subfield to the bundle if dataType is not a null.
+  if (dataType) {
+    Identifier dataId = Identifier::get("data", context);
+    if (isFlip)
+      elements.push_back(BundleElement(dataId, FlipType::get(dataType)));
+    else
+      elements.push_back(BundleElement(dataId, dataType));
+  }
+
+  return BundleType::get(ArrayRef<BundleElement>(elements), context);
+}
+
+/// Return a FIRRTL bundle type (with data, valid, and ready subfields) given a
+/// standard data type. Current supported data types are integer (signed,
+/// unsigned, and signless), index, and none.
+FIRRTLType getBundleType(Type type, bool isFlip) {
```

Comment:
Random suggestion: this method and buildBundleType would be more ergonomic if you pass in. the builder from the client.  Both ways are fine though.